Adds support for OpenAI chat completions API in bench_serving

JustinTong0323 opened the pull request:
<!-- Thank you for your contribution! We appreciate it. The following guidelines will help improve your pull request and facilitate feedback. If anything is unclear, don't hesitate to submit your pull request and ask the maintainers for assistance. -->

## Motivation

<!-- Explain the purpose of this PR and the goals it aims to achieve. -->

## Modifications

Implements a new function `async_request_openai_chat_completions` to support benchmarking OpenAI's chat completions API.

This includes handling both streaming and non-streaming responses, image data within messages, and adapting to the chat API's specific response format.

Also adds corresponding entries for vllm and lmdeploy chat endpoints to `ASYNC_REQUEST_FUNCS`.

Adds option to disable chat template application in mmmu requests to resolve errors with some tokenizers.

<!-- Describe the changes made in this PR. -->

## Checklist

- [x] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [ ] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [ ] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [ ] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [ ] Please feel free to join our Slack channel at https://slack.sglang.ai to discuss your PR.

JustinTong0323 committed:
Adds support for OpenAI chat completions API in bench_serving

Implements a new function `async_request_openai_chat_completions`
to support benchmarking OpenAI's chat completions API.

This includes handling both streaming and non-streaming responses,
image data within messages, and adapting to the chat API's specific
response format.

Also adds corresponding entries for vllm and lmdeploy chat
endpoints to `ASYNC_REQUEST_FUNCS`.

Adds option to disable chat template application in mmmu requests to resolve errors with some tokenizers.

Signed-off-by: Xinyuan Tong <justinning0323@outlook.com>

JustinTong0323 committed:
Merge branch 'main' into bench_serving_openai_chat

yhyang201 commented:
While you're at it, could you also change `0.0.0.0` to `127.0.0.1` on line 1797? Thanks!

yhyang201 committed:
Merge branch 'main' into bench_serving_openai_chat

JustinTong0323 commented:
> While you're at it, could you also change `0.0.0.0` to `127.0.0.1` on line 1797? Thanks!

I am not sure about the intent here. `0.0.0.0` seems a more general port as default. What's your concern about it?

mickqian commented on python/sglang/bench_serving.py:
```diff
@@ -555,8 +674,11 @@ def get_dataset(args, tokenizer):
     "sglang": async_request_sglang_generate,
     "sglang-native": async_request_sglang_generate,
     "sglang-oai": async_request_openai_completions,
+    "sglang-oai-chat": async_request_openai_chat_completions,
```
Can you describe the major difference between `sglang-oai-chat` and `sglang`, especially for vlm benchmarking scenarios? Thanks

CatherineSue commented on python/sglang/bench_serving.py:
```diff
@@ -661,6 +783,7 @@ def sample_mmmu_requests(
     num_requests: int,
     tokenizer: PreTrainedTokenizerBase,
     fixed_output_len: Optional[int] = None,
+    apply_chat_template: bool = True,
```
+1 I think the docstring is nice to have

CatherineSue reviewed (commented):
Left some small comments

JustinTong0323 committed:
Merge branch 'main' into bench_serving_openai_chat

JustinTong0323 committed:
Document and clarify OpenAI chat completions request

Adds a detailed docstring to the OpenAI chat completions request function, explaining its functionality, arguments, and return values.

Also adds a description for the `apply_chat_template` argument in `sample_mmmu_requests`.

Signed-off-by: Xinyuan Tong <justinning0323@outlook.com>

JustinTong0323 committed:
Merge branch 'main' into bench_serving_openai_chat

JustinTong0323 committed:
Merge branch 'main' into bench_serving_openai_chat

JustinTong0323 committed:
Merge branch 'main' into bench_serving_openai_chat

JustinTong0323 committed:
Merge branch 'main' into bench_serving_openai_chat

mickqian committed:
Merge branch 'main' into bench_serving_openai_chat

JustinTong0323 committed:
Merge branch 'main' into bench_serving_openai_chat

JustinTong0323 committed:
Fix: Minicpm

Fixes an issue where the `WHISPER_ATTENTION_CLASSES` dictionary
was being used to access attention classes, which is not
necessary. This commit directly uses `WhisperAttention` to
initialize the attention layer, simplifying the code and
potentially improving performance.

Signed-off-by: Xinyuan Tong <justinning0323@outlook.com>

JustinTong0323 committed:
Merge branch 'main' into bench_serving_openai_chat

merrymercy commented on python/sglang/bench_serving.py:
```diff
@@ -1538,12 +1678,19 @@ def run_benchmark(args_: argparse.Namespace):
             if args.base_url
             else f"http://{args.host}:{args.port}/generate"
         )
+        args.apply_chat_template = True
```
remove this debug code
